src/execution/operator/persistent/physical_insert.cpp

```diff
@@ -108,14 +121,31 @@ unique_ptr<GlobalSourceState> PhysicalInsert::GetGlobalSourceState(ClientContext
 void PhysicalInsert::GetData(ExecutionContext &context, DataChunk &chunk, GlobalSourceState &gstate,
                              LocalSourceState &lstate) const {
 	auto &state = (InsertSourceState &)gstate;
-	auto &g = (InsertGlobalState &)*sink_state;
+	auto &insert_gstate = (InsertGlobalState &)*sink_state;
 	if (state.finished) {
 		return;
 	}
-
-	chunk.SetCardinality(1);
-	chunk.SetValue(0, 0, Value::BIGINT(g.insert_count));
-	state.finished = true;
+	if (!return_chunk) {
+		chunk.SetCardinality(1);
+		chunk.SetValue(0, 0, Value::BIGINT(insert_gstate.insert_count));
+		state.finished = true;
+	} else {
+		idx_t chunk_return = insert_gstate.returned_chunk_count;
+
+		if (insert_gstate.return_chunk_collection.Chunks().size() > chunk_return) {
```

Comment:
This code can be somewhat simplified: we can just return if `chunk_return` exceeds the number of available chunks:

```cpp
if (chunk_return >= insert_gstate.return_chunk_collection.Chunks().size()) {
    return;
}
```

Setting `state.finished` is not necessary in this case, neither is resetting the chunk.